addons/binding/org.openhab.binding.s7/src/main/java/org/openhab/binding/s7/handler/S7BridgeHandler.java

```diff
@@ -0,0 +1,342 @@
+
+    private void createClient() {
+        if (getConfig().get(HOST) != null) {
+            if (client == null) {
+                client = new S7Client();
+                connectClient();
+            }
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR,
+                    "Cannot connect to S7 Server. IP address not set.");
+        }
+    }
+
+    private boolean connectClient() {
+        if (client.Connected) {
+            return true;
+        }
+
+        String hostname = (String) getConfig().get(HOST);
+        int localTSAP = ((java.math.BigDecimal) getConfig().get(LOCAL_TSAP)).intValue();
+        int remoteTSAP = ((java.math.BigDecimal) getConfig().get(REMOTE_TSAP)).intValue();
+
+        client.SetConnectionParams(hostname, localTSAP, remoteTSAP);
+
+        client.Connect();
+
+        if (client.Connected) {
+            onConnectionEstablished();
+        }
+
+        return client.Connected;
+    }
+
+    private synchronized void onUpdate() {
+        destroyClient();
+        createClient();
+
+        destroyPollingJob();
+        createPollingJob();
+    }
+
+    private void destroyPollingJob() {
+        if (pollingJob != null) {
+            pollingJob.cancel(false);
+
+            while (!pollingJob.isDone()) {
+                try {
+                    Thread.sleep(5);
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                    break;
+                }
+            }
+
+            pollingJob = null;
+        }
+    }
+
+    private void destroyClient() {
+        if (client != null) {
+            client.Disconnect();
+            client = null;
+        }
+    }
+
+    private void createPollingJob() {
+        logger.debug("Creating new polling job for S7 server {} with interval {}ms.", getConfig().get(HOST),
+                getConfig().get(POLLING_INTERVAL));
+        int pollingInterval = DEFAULT_POLLING_INTERVAL;
+        try {
+            Object pollingIntervalConfig = getConfig().get(POLLING_INTERVAL);
+            if (pollingIntervalConfig != null) {
+                pollingInterval = ((BigDecimal) pollingIntervalConfig).intValue();
+            } else {
+                logger.debug("Polling interval not configured for this S7 client. Using default value: {}s",
+                        pollingInterval);
+            }
+        } catch (NumberFormatException ex) {
+            logger.info("Wrong configuration value for polling interval. Using default value: {}s", pollingInterval);
+        }
+
+        pollingJob = scheduler.scheduleWithFixedDelay(pollingRunnable, 5000, pollingInterval, TimeUnit.MILLISECONDS);
+    }
+
+    private Hashtable<Integer, byte[]> ReadData() {
+        final int PALength = 64;
+        final int PELength = 64;
+        final int DBLength = 96;
+        final int MKLength = 32;
+
+        Hashtable<Integer, byte[]> data = new Hashtable<Integer, byte[]>();
```

Comment:
Hash table is old fashioned and relatively slow please use an Map for the type declaration and an hashmap as the implementation.